lib/src/default_index/revset_engine.rs

```diff
@@ -927,15 +930,21 @@ impl EvaluationContext<'_> {
                 let sources_revset = self.evaluate(sources)?;
                 let mut sources_predicate = sources_revset.to_predicate_fn();
                 let mut set_reps = HashSet::new();
-                for &pos in &domain_vec {
+                for (&pos, &rep) in domain_vec.iter().zip(&domain_reps) {
+                    // Skip evaluating predicate if `rep` has already been added.
+                    if set_reps.contains(&rep) {
+                        continue;
+                    }
                     if sources_predicate(index, pos)? {
-                        set_reps.insert(sets.find(pos));
+                        set_reps.insert(rep);
                     }
                 }
 
                 let positions = domain_vec
                     .into_iter()
-                    .filter(|pos| set_reps.contains(&sets.find(*pos)))
+                    .zip(domain_reps)
+                    .filter(|(_, rep)| set_reps.contains(rep))
+                    .map(|(domain, _)| domain)
```

Comment:
nit: maybe `position` is a better variable name so it doesn't sound like it's the whole domain